smart-contracts/wasm-chain-integration/src/v1/trie/api.rs

```diff
@@ -146,17 +149,48 @@ impl PersistentState {
         }
     }
 
-    /// Lookup a key in the tree. This is only meant for testing
-    /// since performance is slow compared to lookup in the mutable tree.
+    /// Lookup a key in the tree and return a copy of the value stored.
     pub fn lookup(&self, loader: &mut impl BackingStoreLoad, key: &[u8]) -> Option<Value> {
         match self {
             PersistentState::Empty => None,
-            PersistentState::Root(node) => {
-                let node = node.get(loader);
-                let data = node.data.lookup(loader, key)?;
-                let borrowed = data.borrow();
-                Some(borrowed.get_copy(loader))
+            PersistentState::Root(root_node) => {
+                let mut trie = root_node.make_mutable(0, loader);
+                let entry = trie.get_entry(loader, key)?;
+                trie.with_entry(entry, loader, |x| x.to_vec())
+            }
+        }
+    }
+
+    /// Derive a fresh mutable trie from the [`PersistentState`] using the given
+    /// loader. In contrast to using [`thaw`](Self::thaw) and then using
+    /// [`MutableState::get_inner`] this directly yields a mutable trie that
+    /// can be used. However this mutable trie is not shareable and does not
+    /// have checkpointing in contrast to the [`MutableState`].
+    ///
+    /// This is intended when the persistent trie is used in read-only way. It
+    /// gives access to the efficient implementations of lookup and
```

Comment:
Is it correct that `into_trie` only should be preferred over `lookup` if you need to look up multiple values? If so, then we should add a comment about that.

Reply:
Yes, good point, will add.